File: Anya_Seth/BinarySearch/Special Array With X Elements Greater Than or Equal X.java
class Solution {
    public int specialArray(int[] nums) {
     int n = nums.length;
        int low = 0;
        int high = n;

        while (low <= high) {
            int mid = (low + high) / 2;
            if (isPossible(nums, mid) == mid) {
                return mid;
            } else if (isPossible(nums, mid) > mid) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }

    private int isPossible(int[] nums, int n) {
        int count = 0;
        for (int num : nums) {
            if (num >= n) {
                count++;
            }
        }
        return count;
    }
}

File: NikhilKumarGupta/Binary search/Special Array With X Elements Greater Than or Equal X.cpp
class Solution {
    public:
    int count(vector<int>&nums,int mid){
        
        int n=nums.size();
        int cnt=0;
        for(int i=0;i<n;i++){
            if(nums[i]>=mid){
                cnt++;
            }
        }
        return cnt;
    }
        int specialArray(vector<int>& nums) {
            
            int n=nums.size();
            int low=0;
            int high=n;
            while(low<=high){
                int mid=(low+high)/2;
                if(count(nums,mid)==mid){
                    return mid;
                   
                }
                else if( count(nums,mid)>mid ){
                     low=mid+1;
                }
                else {
                    high=mid-1;
                }
                
    
            }
            return -1;
    
    
            
        }
    };